### DISCUSSION TITLE
bug? Global Dependencies is no effective

### DISCUSSION THREAD

**Discussion by un-human** [fastapi/fastapi#6914] [Questions] [answered]
### bug?
app Global Dependencies is no effective

### env
python==3.7.8
fastapi==0.61.1
windows10

### coding

- api_dosc : https://fastapi.tiangolo.com/zh/tutorial/dependencies/global-dependencies/

```python
from fastapi import Depends, FastAPI, Header, HTTPException


async def verify_token(x_token: str = Header(...)):
    if x_token != "fake-super-secret-token":
        raise HTTPException(status_code=400, detail="X-Token header invalid")


async def verify_key(x_key: str = Header(...)):
    if x_key != "fake-super-secret-key":
        raise HTTPException(status_code=400, detail="X-Key header invalid")
    return x_key



app = FastAPI(dependencies=[Depends(verify_token), Depends(verify_key)])



@app.get("/items/")
async def read_items():
    return [{"item": "Portal Gun"}, {"item": "Plumbus"}]


@app.get("/users/")
async def read_users():
    return [{"username": "Rick"}, {"username": "Morty"}]

if __name__ == '__main__':
    uvicorn.run(app, host='0.0.0.0', port=8000, reload=False, debug=True, workers=1)
```

**Comment by Mause** [comment]
Your code seems to work as expected. What do you think is not working?

**Comment by un-human** [comment]
> Your code seems to work as expected. What do you think is not working?

 Header(...)  no  x_token and no verify_key
curl /items/  get response not   raise HTTPException

**Comment by Mause** [comment]
Well yes, you've declared them (via the ellipses `...`) as required, so if you don't supply them, validation will fail. So your own validation code won't run

**Comment by un-human** [comment]
> Well yes, you've declared them (via the ellipses `...`) as required, so if you don't supply them, validation will fail. So your own validation code won't run

I mean：
curl   /items/   with Header has no x_token  and x_key
but the result is igot the response  and  status_code==200, it is not normal ;
The normal outcome Should  be raise HTTPException（Global Dependencies is no effective ）

**Comment by insomnes** [comment]
@stringk245 
Can you show 200 code? I get 422 as it should be: as @Mause mentioned you marked token and key as required, and FastAPI catches the problem with no required header in request as mentioned:
```
http --follow :8000/items/
HTTP/1.1 422 Unprocessable Entity
content-length: 172
content-type: application/json
date: Thu, 11 Mar 2021 07:15:02 GMT
server: uvicorn

{
    "detail": [
        {
            "loc": [
                "header",
                "x-token"
            ],
            "msg": "field required",
            "type": "value_error.missing"
        },
        {
            "loc": [
                "header",
                "x-key"
            ],
            "msg": "field required",
            "type": "value_error.missing"
        }
    ]
}
```


You can do something like that, for getting your error (setup default `None` value):
```python3
from fastapi import Depends, FastAPI, Header, HTTPException
from typing import Optional


async def verify_token(x_token: Optional[str] = Header(None)):
    if x_token != "fake-super-secret-token":
        raise HTTPException(status_code=400, detail="X-Token header invalid")


async def verify_key(x_key: Optional[str] = Header(None)):
    if x_key != "fake-super-secret-key":
        raise HTTPException(status_code=400, detail="X-Key header invalid")
    return x_key


app = FastAPI(dependencies=[Depends(verify_token), Depends(verify_key)])


@app.get("/items/")
async def read_items():
    return [{"item": "Portal Gun"}, {"item": "Plumbus"}]


@app.get("/users/")
async def read_users():
    return [{"username": "Rick"}, {"username": "Morty"}]
```
==>
```
 http --follow :8000/items/
HTTP/1.1 400 Bad Request
content-length: 35
content-type: application/json
date: Thu, 11 Mar 2021 07:19:23 GMT
server: uvicorn

{
    "detail": "X-Token header invalid"
}
```

**Comment by un-human** [comment]
> marked

see the img ↓
![image](https://user-images.githubusercontent.com/11296126/110751904-7a585980-827f-11eb-951f-2ae8bac2e872.png)

**Comment by un-human** [comment]
> > marked
> 
> see the img ↓
> ![image](https://user-images.githubusercontent.com/11296126/110751904-7a585980-827f-11eb-951f-2ae8bac2e872.png)

![image](https://user-images.githubusercontent.com/11296126/110752360-2437e600-8280-11eb-82ff-1e9e88c97ddf.png)

**Comment by insomnes** [comment]
@stringk245 
Ok, now i got it. Problem is your version `fastapi==0.61.1`, cause global dependencies were introduced in 0.62.0 release
https://github.com/tiangolo/fastapi/releases/tag/0.62.0

**Comment by insomnes** [comment, marked as answer]
But you should still setup default None value, cause in situation with no headers sent:
```python3
import requests

domain = "http://localhost:8000"


def get_items():
    url = domain + '/items/'
    resp = requests.get(url)
    print(resp.status_code)
    print(resp.text)


if __name__ == '__main__':
    get_items()
```

You will still get 422:
```
422
{"detail":[{"loc":["header","x-token"],"msg":"field required","type":"value_error.missing"},{"loc":["header","x-key"],"msg":"field required","type":"value_error.missing"}]}
```

**Comment by un-human** [comment]
> @stringk245
> Ok, now i got it. Problem is your version `fastapi==0.61.1`, cause global dependencies were introduced in 0.62.0 release
> https://github.com/tiangolo/fastapi/releases/tag/0.62.0

Thank you very much，i got it，haha

**Comment by insomnes** [comment]
@stringk245 
Can you close the issue if your problem is solved?

**Comment by un-human** [comment]
> But you should still setup default None value, cause in situation with no headers sent:
> 
> ```python
> import requests
> 
> domain = "http://localhost:8000"
> 
> 
> def get_items():
>     url = domain + '/items/'
>     resp = requests.get(url)
>     print(resp.status_code)
>     print(resp.text)
> 
> 
> if __name__ == '__main__':
>     get_items()
> ```
> 
> You will still get 422:
> 
> ```
> 422
> {"detail":[{"loc":["header","x-token"],"msg":"field required","type":"value_error.missing"},{"loc":["header","x-key"],"msg":"field required","type":"value_error.missing"}]}
> ```

Thank you again, i know ,i  do  this  just for test ；
coding-dev  should be user Header(None)  ，avoid client no set Header  get error

**Comment by tiangolo** [comment]
Thanks for the help here @Mause and @stringk245 ! 👏  🙇 

Thanks for reporting back and closing the issue @insomnes  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
